### PR TITLE
Add check for DISTINCT on non-aggregation function

### LINKED ISSUE #1204: Bad error message for DISTINCT on non-aggregation function
This is failing with a `VerifyException` rather than a proper analysis error.

```
presto> SELECT lower(DISTINCT id) FROM (VALUES ('foo')) AS t (id);
Query 20190729_060454_00001_d9dzy failed: window does not support distinct

com.google.common.base.VerifyException: window does not support distinct
	at com.google.common.base.Verify.verify(Verify.java:124)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitFunctionCall(ExpressionInterpreter.java:934)
	at io.prestosql.sql.tree.FunctionCall.accept(FunctionCall.java:110)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:287)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.rewrite(SimplifyExpressions.java:54)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.lambda$createRewrite$0(SimplifyExpressions.java:79)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$ProjectExpressionRewrite.lambda$apply$0(ExpressionRewriteRuleSet.java:125)
	at io.prestosql.sql.planner.plan.Assignments.lambda$rewrite$1(Assignments.java:108)
```

Based on the original report from @yuokada, 9f78af6fec9441532a9a19a0d235e322455588c9 seems to have added validation for this, and previously the bogus keyword was silently ignored.

### PR DESCRIPTION
**Pull request by j4ckofalltrades**
Fail with a proper analysis error with a more appropriate error message
rather than a generic VerifyException.

Fixes #1204

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -944,6 +945,10 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
                 windowFunctions.add(NodeRef.of(node));
             }
 
+            if (node.isDistinct() && extractAggregateFunctions(List.of(node), metadata).isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, node, "DISTINCT is not supported for non-aggregation functions");
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:949]: > `extractAggregateFunctions(List.of(node), metadata).isEmpty()`

This is not specific enough. For example, it won't catch `lower(DISTINCT max(a))`. It should directly check whether the function represented by this FunctionCall is DISTINCT and not an aggregation function.

Also, the check should be done only when the FunctionCall node is not a window function, as that check is handled elsewhere (e.g., `max(DISTINCT a) OVER ()`)

**Review by martint — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (j4ckofalltrades) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: Changes applied, thanks.

**Comment by martint** [comment]
There are some build failures:

```
Error:  src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:[163,15] (imports) UnusedImports: Unused import - io.prestosql.sql.analyzer.ExpressionTreeUtils.extractAggregateFunctions.
Error:  src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:[947,13] (blocks) RightCurly: '}' at column 13 should be alone on a line.
```

**Comment by j4ckofalltrades** [comment]
ci / pt (config-default, suite-6-non-generic) (pull_request) step failed with the following error (then stalled until timeout)?

```
2020-10-06T04:34:36.373Z	ERROR	container-stats-1	io.prestosql.tests.product.launcher.env.StatisticsFetcher	Could not fetch container hadoop-master statistics: java.lang.RuntimeException: java.lang.InterruptedException
```

**Comment by kokosing** [comment]
Test failure seems to be related to: https://github.com/prestosql/presto/issues/5422

**Comment by j4ckofalltrades** [comment]
Build checks pass now.

**Comment by martint** [comment]
Merged, thanks!

**Comment by j4ckofalltrades** [comment]
@martint Looks like to forgot to merge? :)

**Comment by martint** [comment]
No, i merged it directly after squashing the commits, but the github UI doesn’t show it: https://github.com/prestosql/presto/commit/b45c5741cb77012e5cf4d5fee34fe423e206e954

**Comment by j4ckofalltrades** [comment]
Ahh cool, thanks.

**Pull request closed without merging**